New lesson

AleksandrovAleksey opened the pull request:
На скрине всё ещё очень много логики прописано
Начал выносить элементы в отдельные функции

AleksandrovAleksey committed:
add new screen

AleksandrovAleksey committed:
add new view model

AleksandrovAleksey committed:
Rename .java to .kt

AleksandrovAleksey committed:
add repository, state and api
add DaysOfWeekGrid
add try catch in viewModel
add fun ValidateDates

chiitach commented on app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/ui/newLesson/NewLessonScreen.kt:
```diff
@@ -0,0 +1,512 @@
+
+        Box(Modifier.fillMaxSize()) {
+            Column {
+                Row(
```
Если внутри Row ничего больше не будет, можно текст вытащить из Row. Ниже аналогичная ситуация

chiitach commented on app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/ui/newLesson/NewLessonScreen.kt:
```diff
@@ -0,0 +1,512 @@
+    val mYear: Int
+    val mMonth: Int
+    val mDay: Int
+    val mCalendar = Calendar.getInstance()
```
Можно без префикса 'm' в названии переменных. Они путают и в целом нпонятно, зачем нужны

AleksandrovAleksey committed:
Merge branch 'main' into new_lesson

# Conflicts:
#	app/src/main/res/values/strings.xml

AleksandrovAleksey committed:
made and add events for newLesson
made POST method for creating lesson
add new screen, viewModel and state for new assignments
also made string values and events for assignment screen
deleted room from gradle
add Uuid for id of Lesson

chiitach commented on app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/ui/newAssignment/NewAssignmentState.kt:
```diff
@@ -0,0 +1,34 @@
+    val isFormValid: Boolean = false,
+)
+
+sealed interface NewAssignmentEvent {
```
вынести в отдельный файл

chiitach commented on app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/ui/newAssignment/NewAssignmentState.kt:
```diff
@@ -0,0 +1,34 @@
+data class NewAssignmentState(
+    val disciplines: List<Discipline>,
+    val students: List<Student>,
+    val profile: Profile,
```
тут надо использовать ProfileDomain, потому что Profile с data слоя неконсистентный.

chiitach replied:
Напомните, этот экран в принципе только у учителя есть, да?

AleksandrovAleksey replied:
Да, только у учителя должен быть экран

chiitach commented on app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/ui/newAssignment/NewAssignmentScreen.kt:
```diff
@@ -0,0 +1,460 @@
+    viewModel: NewAssignmentViewModel = koinViewModel()
+) {
+    val state by viewModel.stateFlow.collectAsState()
+    val isTeacher = state.profile.teacher != null
```
Если исп. ProfileDomain, тогда тут будет красивая проверка на учителя:
`state.profile.personType is PersonType.TeacherType`

chiitach commented on app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/ui/newAssignment/NewAssignmentViewModel.kt:
```diff
@@ -0,0 +1,219 @@
+        )
+    )
+
+    val stateFlow = _stateFlow.asStateFlow()
```
можно кстати чуть упростить ViewModel, а конкретно методы с обновлением значения и чтобы везде не вызывать метод isFormValid сделать это тут во время получения обновлённого значения. Будет что-то типа:

```
private val initState = NewAssignmentState(
            emptyList(),
            emptyList(),
            profile = baseTeacherProfile,
            selectedStudent = null,
            selectedDiscipline = null,
            dueDate = currentDateTime.date,
            dueTime = currentDateTime.time,
            title = "",
            description = "",
            error = null,
            isLoading = false,
            isFormValid = false,
        )

private val _stateFlow: MutableStateFlow<NewAssignmentState> = MutableStateFlow(initState)

val stateFlow: StateFlow<NewAssignmentState> = _stateFlow.map { state -> 
      state.copy(isFormValid = isFormValid(....)) 
} .stateIn(viewModelScope, SharingStarted.Eagerly, initState)
```

Тогда каждый раз при получении нового значения через апдейты, на юай прокинется значение флоу, которое сделает дополнительный map и заполнит поле isFormValid актуальными данными. И из всех методов, обновляющих поля можно будет просто удалить такие участки кода: 
```
isFormValid = isFormValid(
                    dueDate = currentState.dueDate,
                    dueTime = time,
                    student = currentState.selectedStudent,
                    discipline = currentState.selectedDiscipline,
                    title = currentState.title,
                    description = currentState.description
                )
```

Комментарий больше FYI, на твоё усмотрение исправлять или нет.

chiitach commented on app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/ui/newLesson/NewLessonState.kt:
```diff
@@ -0,0 +1,36 @@
+    val endDate: LocalDate,
+)
+
+sealed interface NewLessonEvent {
```
тоже лучше вынести в другой файл. Потому что стейт экрана и данные ивенты - это разные штуки

chiitach commented on app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/data/LessonRepository.kt:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    suspend fun getDisciplines(): List<Discipline> {
+        return lessonApi.getDiscipline().body() ?: emptyList()
```
```suggestion
        // return lessonApi.getDiscipline().body() ?: emptyList()
        return listOf(Discipline(
        .......
        ))
```

AleksandrovAleksey committed:
Mocked data from repository
Add NewLesson and NewAssignment Events

AleksandrovAleksey committed:
Merge branch 'main' into new_lesson

AleksandrovAleksey committed:
Add switching to NewAssignmentScreen and NewLessonScreen
Some changes in screens for correct view

AleksandrovAleksey committed:
Merge branch 'main' into new_lesson

# Conflicts:
#	app/src/main/java/ru/bmstu/marksUpTeam/android/marksUpApp/ui/schedule/ScheduleScreen.kt
#	build.gradle.kts
#	gradle/wrapper/gradle-wrapper.properties

AleksandrovAleksey committed:
updated branches with new changes from others
fixed problem with profile - got the profileDomain
